magento/magento2
# Fix #36384: IdListBuilder object reuse causes accumulated joins

### Description (*)
When async grid update is enabled the instance of IdListBuilder gets reused across multiple grid update tasks. This causes the associated join to the archived grid tables to accumulate for each and results in joining unrelated archived grid tables.

This is because `IdListBuilder` accumulates joins and results in the following query:
```sql
SELECT `sales_invoice`.`entity_id` FROM `sales_invoice` 
LEFT JOIN `magento_sales_order_grid_archive` ON sales_invoice.entity_id = magento_sales_order_grid_archive.entity_id 
LEFT JOIN `magento_sales_creditmemo_grid_archive` ON sales_invoice.entity_id = magento_sales_creditmemo_grid_archive.entity_id 
LEFT JOIN `magento_sales_invoice_grid_archive` ON sales_invoice.entity_id = magento_sales_invoice_grid_archive.entity_id 
WHERE (sales_invoice.updated_at >= '2023-07-06 05:17:06') 
AND (magento_sales_order_grid_archive.entity_id IS NULL) 
AND (magento_sales_creditmemo_grid_archive.entity_id IS NULL) 
AND (magento_sales_invoice_grid_archive.entity_id IS NULL)
```

In certain circumstances this results in invoices, shipments and creditmemos incorrectly being excluded from the list of IDs not present in the grid.

This change updates the `UpdatedIdListProvider` constructor to create a new instance of `IdListBuilder` instead of reusing existing ones, therefore it doesn't incorrectly share joins across tasks.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#36384

### Manual testing scenarios (*)
1. Ensure `dev/grid/async_indexing` is on
2. Ensure orders marked with `cancelled` state are moved to `magento_sales_order_grid_archive`
3. Create 3 unpaid orders (ensure entity_id = 1, 2, 3)
4. Ensure `sales_order_grid` is updated first
5. Cancel unpaid order entity_id = 1
6. Ensure the cancelled order is moved to `magento_sales_order_grid_archive`
7. Invoice order entity_id = 2 (invoice entity_id should be 1)
8. Note that `sales_invoice_grid` does not receive the new invoice

### Contribution checklist (*)
